pinot-core/src/main/java/org/apache/pinot/core/data/table/OrderedIndexedTableResizer.java

```diff
@@ -0,0 +1,236 @@
+
+      for (int i = 0; i < _numOrderBy; i++) {
+        int result = comparators[i].compare(o1.getValues()[i], o2.getValues()[i]);
+        if (result != 0) {
+          return result;
+        }
+      }
+      return 0;
+    };
+  }
+
+  /**
+   * Constructs an IntermediateRecord from Record
+   * The IntermediateRecord::key is the same Record::key
+   * The IntermediateRecord::values contains only the order by columns, in the query's sort sequence
+   * For aggregation values in the order by, the final result is extracted if the intermediate result is non-comparable
+   */
+  @VisibleForTesting
+  IntermediateRecord getIntermediateRecord(Record record) {
+    Comparable[] intermediateRecordValues = new Comparable[_numOrderBy];
+
+    Object[] keyColumns = record.getKey().getColumns();
+    Object[] aggregationColumns = record.getValues();
+    for (int i = 0; i < _numOrderBy; i++) {
+      OrderByInfo orderByInfo = _orderByInfos.get(i);
+      Comparable comparable;
+      if (orderByInfo.getOrderByType().equals(OrderByType.KEY_COLUMN)) {
+        Object keyColumn = keyColumns[orderByInfo.getIndex()];
+        comparable = (Comparable) keyColumn; // FIXME: is this the right way to get Comparable? will it work for BYTES?
+      } else {
+        Object aggregationColumn = aggregationColumns[orderByInfo.getIndex()];
+        AggregationFunction aggregationFunction = orderByInfo.getAggregationFunction();
+        if (!aggregationFunction.isIntermediateResultComparable()) {
+          comparable = aggregationFunction.extractFinalResult(aggregationColumn);
+        } else {
+          comparable = (Comparable) aggregationColumn;
+        }
+      }
+      intermediateRecordValues[i] = comparable;
+    }
+    return new IntermediateRecord(record.getKey(), intermediateRecordValues);
+  }
+
+  /**
+   * Trim recordsMap to trimToSize, based on order by information
+   */
+  @Override
+  void resizeRecordsMap(Map<Key, Record> recordsMap, int trimToSize) {
+
+    // make min heap of elements to evict
+    int heapSize = recordsMap.size() - trimToSize;
+    if (heapSize > 0) {
+      PriorityQueue<IntermediateRecord> minHeap = new PriorityQueue<>(heapSize, _intermediateRecordComparator);
+
+      for (Record record : recordsMap.values()) {
+
+        IntermediateRecord intermediateRecord = getIntermediateRecord(record);
+        if (minHeap.size() < heapSize) {
+          minHeap.offer(intermediateRecord);
+        } else {
+          IntermediateRecord peek = minHeap.peek();
+          if (minHeap.comparator().compare(peek, intermediateRecord) < 0) {
+            minHeap.poll();
+            minHeap.offer(intermediateRecord);
+          }
+        }
+      }
+
+      for (IntermediateRecord evictRecord : minHeap) {
+        recordsMap.remove(evictRecord.getKey());
+      }
+    }
+  }
+
+  /**
+   * Sort the records in the recordsMap according to order by
+   */
+  @Override
+  List<Record> sortRecordsMap(Map<Key, Record> recordsMap) {
```

Comment:
Will there be a case where the caller will call both resize and sort? If so, in that case there will be an additional sorting that can be avoided.

Reply:
+1. In the final step, you should use PriorityQueue to resize and sort

Reply:
Changed it to resize and sort in one go using the same PQ when finishing.